env/io_posix.cc

```diff
@@ -409,13 +370,112 @@ size_t PosixHelper::GetUniqueIdFromFile(int fd, char* id, size_t max_size) {
   return static_cast<size_t>(rid - id);
 }
 #endif
+
+#ifdef OS_LINUX
+void LogicalBufferSizeCache::CacheLogicalBufferSize(
+    const std::set<std::pair<std::string, int>>& directories) {
+  WriteLock lock(&cache_mutex_);
+  for (auto& dir : directories) {
+    cache_.emplace(dir.first, get_logical_buffer_size_(dir.second));
+  }
+}
+
+void LogicalBufferSizeCache::RemoveCachedLogicalBufferSize(
+    const std::unordered_set<std::string>& directories) {
+  WriteLock lock(&cache_mutex_);
+  for (auto& dir : directories) {
+    cache_.erase(dir);
+  }
+}
+
+size_t LogicalBufferSizeCache::GetLogicalBufferSize(const std::string& fname,
+                                                    int fd) {
+  ReadLock lock(&cache_mutex_);
+  std::string dir = fname.substr(0, fname.find_last_of("/"));
+  if (cache_.find(dir) != cache_.end()) {
+    return cache_[dir];
+  }
+  return get_logical_buffer_size_(fd);
```

Comment:
Can we move it outside the read lock?

Reply:
I thought mostly there is no read-write contention, so the critical section for the read lock won't cause performance drop, and leveraging `ReadLock` at the beginning makes the code cleaner.

But I'm ok with taking extra effort to make all critical sections more dense.